File: backend/internal/models/todo.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Todo struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	UserID    primitive.ObjectID `bson:"userId" json:"userId"`
	Title     string             `bson:"title" json:"title"`
	Subject   string             `bson:"subject,omitempty" json:"subject,omitempty"`
	Difficulty string            `bson:"difficulty,omitempty" json:"difficulty,omitempty"`
	QuestionsTarget int          `bson:"questionsTarget,omitempty" json:"questionsTarget,omitempty"`
	Completed bool               `bson:"completed" json:"completed"`
	ScheduledDate *time.Time     `bson:"scheduledDate,omitempty" json:"scheduledDate,omitempty"`
	OriginalScheduledDate *time.Time `bson:"originalScheduledDate,omitempty" json:"originalScheduledDate,omitempty"`
	RescheduledCount int         `bson:"rescheduledCount,omitempty" json:"rescheduledCount,omitempty"`
	DueDate   *time.Time         `bson:"dueDate,omitempty" json:"dueDate,omitempty"`
	// Schedule integration — set when a todo is generated from an AI schedule.
	Source    string             `bson:"source,omitempty" json:"source,omitempty"`       // "schedule" | "" (manual)
	StartTime string             `bson:"startTime,omitempty" json:"startTime,omitempty"` // "08:00"
	EndTime   string             `bson:"endTime,omitempty" json:"endTime,omitempty"`     // "09:00"
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt"`
}

File: src/types/todo.ts
export type TodoDifficulty = 'easy' | 'medium' | 'hard';

export interface Todo {
  id: string;
  userId?: string;
  title: string;
  subject: string;
  difficulty: TodoDifficulty;
  questionsTarget: number;
  completed: boolean;
  scheduledDate?: string;
  dueDate?: string;
  originalScheduledDate?: string;
  rescheduledCount?: number;
  source?: string;
  startTime?: string;
  endTime?: string;
  createdAt?: string;
  updatedAt?: string;
  /** Client-only marker used while a create request is in flight. */
  optimistic?: boolean;
}

export type CreateTodoInput = Pick<Todo, 'title'> &
  Partial<
    Pick<
      Todo,
      'subject' | 'difficulty' | 'questionsTarget' | 'dueDate' | 'scheduledDate'
    >
  >;

export interface UpdateTodoInput {
  title?: string;
  subject?: string;
  difficulty?: string;
  questionsTarget?: number;
  completed?: boolean;
  dueDate?: string;
  scheduledDate?: string;
}
